hugr/src/types.rs

```diff
@@ -115,6 +115,7 @@ pub(crate) fn least_upper_bound(mut tags: impl Iterator<Item = TypeBound>) -> Ty
 
 #[derive(Clone, PartialEq, Debug, Eq, Serialize, Deserialize)]
 #[serde(tag = "s")]
+#[non_exhaustive]
```

Comment:
This is the only one I'm not sure about; would we ever want other variants here?

Reply:
The current `Unit` variant is just an optimisation for a common case.
Users are not expected to match directly on this enum, but rather to use the uniform interface.

We talked at some point about other special cases we may want to add. Those should probably not require a breaking change if they are only performance-related.